### PR TITLE
Add Code Engine backend

### PR DESCRIPTION
**Pull request by sergii-mamedov**
## Description
It is necessary to add support for Code Engine (CE) as a backend for processing datasets. We plan to replace Cloud Function (CF) with CE. The main feature of CE for us is support for up to 32 GB of RAM.
Steps required:
1. Create a Docker image for CE.
2. Change a serverless backend in `config.json`
3. Change code in `executor.py` to support CE.

### Install IBM CLI on web EC2 or local computer

1. Download and install: 
    `curl -sL https://raw.githubusercontent.com/IBM-Cloud/ibm-cloud-developer-tools/master/linux-installer/idt-installer | bash`
2. Install CE plugin: 
    `ibmcloud plugin install code-engine`
3. Log in:
    `ibmcloud login` by email/password or `ibmcloud login --apikey API_KEY`
4. Targeted resource group (staging - `dev-shared`, production - `metaspace-env-prod`): 
     `ibmcloud target -g RESOURCE_GROUP_NAME`
5. Create kubernetis config file (staging - `metaspace-staging-ce`, production - `metaspace-production-ce`):
     `ibmcloud ce project select --name NAME --kubecfg`
6. Show path:
     `ibmcloud ce project current`

### Docker image for the CE
A sample Dockerfile for CE is available in the [lithops repository](https://github.com/lithops-cloud/lithops/blob/master/runtime/code_engine/Dockerfile). Unfortunately, this Dockerfile is highly redundant. I took it as a basis by removing unnecessary dependencies and adding the ones we need (from [CF Dockerifle](https://github.com/metaspace2020/metaspace/blob/master/metaspace/engine/docker/lithops_ibm_cf/Dockerfile)).

### Other changes
For CE, it is mandatory to set the number of CPUs, while there are only [certain ratios](https://cloud.ibm.com/docs/codeengine?topic=codeengine-mem-cpu-combo) between CPU and RAM.

### CODE REVIEW

File: metaspace/engine/docker/lithops_ibm_cf/Dockerfile [hunk 1]
```diff
@@ -1,27 +1,18 @@
...
+        flask==2.0.1 \
+        gevent==21.8.0 \
+        ibm-cos-sdk==2.10.0 \
+        pika==1.2.0 \
         ps-mem==3.12 \
-        tblib==1.7.0 \
-        namegenerator==1.0.6
+        redis==3.5.3 \
+        requests==2.26.0 \
+        tblib==1.7.0
```
Comment (sergii-mamedov) [inline metaspace/engine/docker/lithops_ibm_cf/Dockerfile:15]: sorted the list and removed unused dependencies

File: metaspace/engine/sm/engine/annotation_lithops/executor.py [hunk 2]
```diff
@@ -24,11 +24,13 @@
 #: manually updating their config files every time it changes. The image must be public on
 #: Docker Hub, and can be rebuilt using the scripts/Dockerfile in `engine/docker/lithops_ibm_cf`.
 #: Note: sci-test changes this constant to force local execution without docker
-RUNTIME_DOCKER_IMAGE = 'metaspace2020/metaspace-lithops:1.8.4'
+RUNTIME_CF_VPC = 'metaspace2020/metaspace-lithops:1.8.5'
+RUNTIME_CE = 'metaspace2020/metaspace-lithops-ce:1.8.6.9'
```
Comment (sergii-mamedov) [inline metaspace/engine/sm/engine/annotation_lithops/executor.py:28]: will change for a major version after review

**Review by LachlanStuart — APPROVED** [review]
These changes look good, but we should discuss the deployment plan and wait for feedback from IBM about the latest bug (3 `DiskPressure` failures causing a job to stall) before this is merged.

**Comment by gilv** [comment]
@sergii-mamedov hold on please...you should never create that zip file on your own. i think something is wrong the way you use it... let me review it and let us discuss it then. if you create that zip file manually - something fundamentally wrong the way you use it (@LachlanStuart fyi )

**Comment by sergii-mamedov** [comment]
@gilv I figured it out, thanks. Unfortunately, this was somehow not clearly stated in the documentation.
It is also not very convenient that I need to use `ibm_cos` to create a docker image. In this case, I need to create a temporary config file with all credentials.

**Comment by gilv** [comment]
@sergii-mamedov i don't think you need ibm_cos to create docker runtime image with Lithops. At what point it asked you COS?

**Comment by sergii-mamedov** [comment]
@gilv 
If I run lithops without config file, I see this:
```
$ lithops runtime build metaspace2020/metaspace-lithops-ce:test -b code_engine
2021-08-17 23:42:10,601 [INFO] lithops.config -- Lithops v2.4.1
2021-08-17 23:42:10,601 [DEBUG] lithops.config -- Config not found. Setting Lithops to localhost mode
2021-08-17 23:42:10,601 [DEBUG] lithops.config -- Loading Serverless backend module: code_engine
2021-08-17 23:42:10,873 [DEBUG] lithops.config -- Loading Storage backend module: localhost
Traceback (most recent call last):
  File "/home/finve/Documents/embl_code/metaspace/metaspace/engine/venv/bin/lithops", line 8, in <module>
    sys.exit(lithops_cli())
  File "/home/finve/Documents/embl_code/metaspace/metaspace/engine/venv/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/finve/Documents/embl_code/metaspace/metaspace/engine/venv/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/finve/Documents/embl_code/metaspace/metaspace/engine/venv/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/finve/Documents/embl_code/metaspace/metaspace/engine/venv/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/finve/Documents/embl_code/metaspace/metaspace/engine/venv/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/finve/Documents/embl_code/metaspace/metaspace/engine/venv/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/finve/Documents/embl_code/metaspace/metaspace/engine/venv/lib/python3.8/site-packages/lithops/scripts/cli.py", line 394, in build
    config = default_config(config, config_ow)
  File "/home/finve/Documents/embl_code/metaspace/metaspace/engine/venv/lib/python3.8/site-packages/lithops/config.py", line 239, in default_config
    raise Exception('Localhost storage backend cannot be used in {} mode'.format(mode))
Exception: Localhost storage backend cannot be used in serverless mode
```

if I create local config file `.lithops_config`:
```
lithops: 
    backend: code_engine
    mode: serverless
```
I see this error:
```
2021-08-17 23:45:08,624 [DEBUG] lithops.config -- Loading Serverless backend module: code_engine
2021-08-17 23:45:08,893 [DEBUG] lithops.config -- Loading Storage backend module: ibm_cos
Traceback (most recent call last):
  File "/home/finve/Documents/embl_code/metaspace/metaspace/engine/venv/bin/lithops", line 8, in <module>
    sys.exit(lithops_cli())
  File "/home/finve/Documents/embl_code/metaspace/metaspace/engine/venv/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/finve/Documents/embl_code/metaspace/metaspace/engine/venv/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/finve/Documents/embl_code/metaspace/metaspace/engine/venv/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/finve/Documents/embl_code/metaspace/metaspace/engine/venv/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/finve/Documents/embl_code/metaspace/metaspace/engine/venv/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/finve/Documents/embl_code/metaspace/metaspace/engine/venv/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/finve/Documents/embl_code/metaspace/metaspace/engine/venv/lib/python3.8/site-packages/lithops/scripts/cli.py", line 394, in build
    config = default_config(config, config_ow)
  File "/home/finve/Documents/embl_code/metaspace/metaspace/engine/venv/lib/python3.8/site-packages/lithops/config.py", line 236, in default_config
    config_data = default_storage_config(config_data)
  File "/home/finve/Documents/embl_code/metaspace/metaspace/engine/venv/lib/python3.8/site-packages/lithops/config.py", line 261, in default_storage_config
    sb_config.load_config(config_data)
  File "/home/finve/Documents/embl_code/metaspace/metaspace/engine/venv/lib/python3.8/site-packages/lithops/storage/backends/ibm_cos/config.py", line 24, in load_config
    raise Exception("ibm_cos section is mandatory in the configuration")
Exception: ibm_cos section is mandatory in the configuration
```

**Comment by gilv** [comment]
@sergii-mamedov if you deploy Lithops to CE, then you need COS. But I thought you need to build Lithops runtime, for this i am not sure you need COS, but i can check. But what is the issue, you need COS anyway, no?

**Comment by sergii-mamedov** [comment]
@gilv Imaging and our application are two different processes. We store all keys encrypted for the main application. It turns out that in order to just create an image, I need to decrypt the key store, create a temporary config file and only after that create a docker image. Moreover, after creating the docker image, I do not see any files on the COS.
Of course, all this can be done on our side. But I would like to understand if you really need access to an COS to create a docker image, or if this is a "feature" of the lithops code.

File: metaspace/engine/sm/engine/annotation_lithops/executor.py [hunk 3]
```diff
@@ -142,8 +143,8 @@ class Executor:
     If a feature is successful, it should be upstreamed to Lithops as an RFC or PR.
 
     Current features:
-      * Switch to the Standalone executor if >4GB of memory is required
-      * Retry with 2x more memory if an execution fails due to an OOM
+      * Switch to the Standalone executor if >32GB of memory is required
+      * Retry with 3x more memory if an execution fails due to an OOM
```
Comment (sergii-mamedov) [inline metaspace/engine/sm/engine/annotation_lithops/executor.py:147]: Increased the number of retry, as we have a wider range of available RAM volumes.

**Review by lmacielvieira — APPROVED** [review]

**Pull request merged**